pkg/csi/service/common/authmanager.go

```diff
@@ -135,44 +184,244 @@ func GenerateDatastoreIgnoreMapForBlockVolumes(ctx context.Context, vc *vsphere.
+		return nil, err
+	}
+	log.Debugf("dsToFileServiceEnabledMap is %+v", dsToFileServiceEnabledMap)
+	// Now create a map of datastores which are queried in the method.
+	dsToFSEnabledMapToReturn := make(map[string]bool)
+	for _, datastoreURL := range datastoreUrls {
+		if val, ok := dsToFileServiceEnabledMap[datastoreURL]; ok {
+			if !val {
+				msg := fmt.Sprintf("File service is not enabled on the datastore: %s", datastoreURL)
+				log.Debugf(msg)
+				dsToFSEnabledMapToReturn[datastoreURL] = false
+			} else {
+				msg := fmt.Sprintf("File service is enabled on the datastore: %s", datastoreURL)
+				log.Debugf(msg)
+				dsToFSEnabledMapToReturn[datastoreURL] = true
+			}
+		} else {
+			msg := fmt.Sprintf("File service is not enabled on the datastore: %s", datastoreURL)
+			log.Debugf(msg)
+			dsToFSEnabledMapToReturn[datastoreURL] = false
+		}
+	}
+	return dsToFSEnabledMapToReturn, nil
+}
+
+func getDatastoresWithRequiredPriv(ctx context.Context, vc *vsphere.VirtualCenter,
+	dsURLs []string, dsInfos []*cnsvsphere.DatastoreInfo, entities []types.ManagedObjectReference) (map[string]*cnsvsphere.DatastoreInfo, error) {
+	log := logger.GetLogger(ctx)
+	log.Debugf("auth manager: file - dsURLs %v dsInfos %v", dsURLs, dsInfos)
+	// dsURLToInfoMap will store a list of vSAN datastores with vSAN FS enabled for which
+	// CSI user has privilege to
+	dsURLToInfoMap := make(map[string]*cnsvsphere.DatastoreInfo)
 	// get authMgr
 	authMgr := object.NewAuthorizationManager(vc.Client.Client)
 	privIds := []string{DsPriv, SysReadPriv}
-	var entities []types.ManagedObjectReference
-	for _, dsInfo := range dsInfos {
-		// go through all datastores to build entities
-		dsMoRef := dsInfo.Datastore.Datastore.Reference()
-		entities = append(entities, dsMoRef)
-	}
 
-	var result []types.EntityPrivilege
 	userName := vc.Config.Username
 	// invoke authMgr function HasUserPrivilegeOnEntities
-	result, err = authMgr.HasUserPrivilegeOnEntities(ctx, entities, userName, privIds)
+	result, err := authMgr.HasUserPrivilegeOnEntities(ctx, entities, userName, privIds)
 	if err != nil {
 		log.Errorf("auth manager: failed to check privilege %v on entities %v for user %s", privIds, entities, userName)
 		return nil, err
 	}
 	log.Debugf("auth manager: HasUserPrivilegeOnEntities returns %v when checking privileges %v on entities %v for user %s", result, privIds, entities, userName)
 	for index, entityPriv := range result {
-		needIgnore := false
+		hasPriv := true
 		privAvails := entityPriv.PrivAvailability
 		for _, privAvail := range privAvails {
 			// required privilege is not grant for this entity
 			if !privAvail.IsGranted {
-				needIgnore = true
+				hasPriv = false
 				break
 			}
 		}
-		if needIgnore {
-			dsIgnoreURLToInfoMap[dsURLs[index]] = dsInfos[index]
-			log.Debugf("auth manager: datastore with URL %s  and name %s is added to dsIgnoreURLToInfoMap", dsInfos[index].Info.Name, dsURLs[index])
+		if hasPriv {
+			dsURLToInfoMap[dsURLs[index]] = dsInfos[index]
```

Comment:
Is the logic based on the index at which the result is received and the index in dsURLs and dsInfos? If yes, this is risky once the indexes are out of order.

Reply:
This is logic which @lipingxue has written for block volumes. I just used as is. Since this is existing code, i will create a separate bug and address this in next PR.